File: solutions/go/06/dinic/main.go
package main

import (
	"fmt"
)

type edge struct {
	to, cap, rev int
}

type Dinic struct {
	n   int
	g   [][]edge
}

func NewDinic(n int) *Dinic {
	g := make([][]edge, n)
	return &Dinic{n, g}
}

func (d *Dinic) AddEdge(fr, to, cap int) {
	d.g[fr] = append(d.g[fr], edge{to, cap, len(d.g[to])})
	d.g[to] = append(d.g[to], edge{fr, 0, len(d.g[fr]) - 1})
}

func (d *Dinic) bfs(s, t int, level []int) bool {
	for i := range level {
		level[i] = -1
	}
	level[s] = 0
	q := []int{s}
	for head := 0; head < len(q); head++ {
		v := q[head]
		for _, e := range d.g[v] {
			if e.cap > 0 && level[e.to] < 0 {
				level[e.to] = level[v] + 1
				q = append(q, e.to)
			}
		}
	}
	return level[t] >= 0
}

func (d *Dinic) dfs(v, t, f int, level, iter []int) int {
	if v == t {
		return f
	}
	for i := iter[v]; i < len(d.g[v]); i++ {
		iter[v] = i
		e := d.g[v][i]
		if e.cap > 0 && level[v] < level[e.to] {
			df := d.dfs(e.to, t, min(f, e.cap), level, iter)
			if df > 0 {
				d.g[v][i].cap -= df
				d.g[e.to][e.rev].cap += df
				return df
			}
		}
	}
	return 0
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func (d *Dinic) MaxFlow(s, t int) int {
	flow := 0
	level := make([]int, d.n)
	for d.bfs(s, t, level) {
		iter := make([]int, d.n)
		for {
			f := d.dfs(s, t, 1<<30, level, iter)
			if f == 0 {
				break
			}
			flow += f
		}
	}
	return flow
}

func main() {
	d := NewDinic(4)
	d.AddEdge(0, 1, 3)
	d.AddEdge(0, 2, 2)
	d.AddEdge(1, 2, 1)
	d.AddEdge(1, 3, 2)
	d.AddEdge(2, 3, 3)
	fmt.Println(d.MaxFlow(0, 3))
}

File: solutions/python/06/dinic/main.py
from collections import deque


class Dinic:
    def __init__(self, n: int) -> None:
        self.n = n
        self.g = [[] for _ in range(n)]

    def add_edge(self, fr: int, to: int, cap: int) -> None:
        self.g[fr].append([to, cap, len(self.g[to])])
        self.g[to].append([fr, 0, len(self.g[fr]) - 1])

    def bfs(self, s: int, t: int, level: list) -> bool:
        for i in range(self.n):
            level[i] = -1
        level[s] = 0
        q = deque([s])
        while q:
            v = q.popleft()
            for to, cap, _ in self.g[v]:
                if cap > 0 and level[to] < 0:
                    level[to] = level[v] + 1
                    q.append(to)
        return level[t] >= 0

    def dfs(self, v: int, t: int, f: int, level: list, iter_: list) -> int:
        if v == t:
            return f
        for i in range(iter_[v], len(self.g[v])):
            iter_[v] = i
            to, cap, rev = self.g[v][i]
            if cap > 0 and level[v] < level[to]:
                d = self.dfs(to, t, min(f, cap), level, iter_)
                if d > 0:
                    self.g[v][i][1] -= d
                    self.g[to][rev][1] += d
                    return d
        return 0

    def max_flow(self, s: int, t: int) -> int:
        flow = 0
        level = [-1] * self.n
        while self.bfs(s, t, level):
            iter_ = [0] * self.n
            while True:
                f = self.dfs(s, t, 10**18, level, iter_)
                if f == 0:
                    break
                flow += f
        return flow


def main() -> None:
    n = 4
    dinic = Dinic(n)
    dinic.add_edge(0, 1, 3)
    dinic.add_edge(0, 2, 2)
    dinic.add_edge(1, 2, 1)
    dinic.add_edge(1, 3, 2)
    dinic.add_edge(2, 3, 3)
    print(dinic.max_flow(0, 3))


if __name__ == "__main__":
    main()
